prefect_kubernetes/credentials.py

```diff
@@ -0,0 +1,133 @@
+            )
+        ```
+    """
+
+    _block_type_name = "Kubernetes Credentials"
+    _logo_url = "https://images.ctfassets.net/zscdif0zqppk/oYuHjIbc26oilfQSEMjRv/a61f5f6ef406eead2df5231835b4c4c2/logo.png?h=250"  # noqa
+
+    cluster_config: Optional[KubernetesClusterConfig] = None
+
+    @contextmanager
+    def get_app_client(self) -> Generator[AppsV1Api, None, None]:
+        """Convenience method for retrieving a kubernetes api client for deployment resources
+
+        Returns:
+            Kubernetes api client generator to interact with "deployment" resources.
+        """
+        generic_client = self.get_generic_client()
+        try:
+            yield AppsV1Api(api_client=generic_client)
+        finally:
+            generic_client.rest_client.pool_manager.clear()
+
+    @contextmanager
+    def get_batch_client(self) -> Generator[BatchV1Api, None, None]:
+        """Convenience method for retrieving a kubernetes api client for job resources.
+
+        Returns:
+            Kubernetes api client generator to interact with "job" resources.
+        """
+        generic_client = self.get_generic_client()
+        try:
+            yield BatchV1Api(api_client=generic_client)
+        finally:
+            generic_client.rest_client.pool_manager.clear()
+
+    @contextmanager
+    def get_core_client(self) -> Generator[CoreV1Api, None, None]:
+        """Convenience method for retrieving a kubernetes api client for core resources.
+
+        Returns:
+            Kubernetes api client generator to interact with "pod", "service"
+            and "secret" resources.
+        """
+        generic_client = self.get_generic_client()
+        try:
+            yield CoreV1Api(api_client=generic_client)
+        finally:
+            generic_client.rest_client.pool_manager.clear()
+
+    def get_generic_client(self) -> ApiClient:
+        """
+        Utility function for configuring a generic Kubernetes client.
+        It will attempt to connect to a Kubernetes cluster in three steps with
+        the first successful connection attempt becoming the mode of communication with
+        a cluster:
+
+        1. It will first attempt to use a `KubernetesCredentials` block's
+        `cluster_config` to configure a client using
+        `KubernetesClusterConfig.configure_client`.
+
+        2. Attempt in-cluster connection (will only work when running on a pod).
+
+        3. Attempt out-of-cluster connection using the default location for a
+        kube config file.
+
+        Returns:
+            An authenticated, generic Kubernetes Client.
+        """
+
+        with ApiClient() as client:
+            if self.cluster_config:
+                self.cluster_config.configure_client()
+                return client
+            else:
+                try:
+                    kube_config.load_incluster_config()
+                except ConfigException:
+                    kube_config.load_kube_config()
+
+                return client
```

Comment:
Based on  the call order of: https://github.com/PrefectHQ/prefect/blob/2da92c03ba917aebc7897b2828c5a9f575c131e9/src/prefect/infrastructure/kubernetes.py#L266-L277,

I think we should do:
1. run all the configs
2. return ApiClient()
(ensure the indentation is correct below since I'm editing on GH)
```suggestion
        if self.cluster_config:
            self.cluster_config.configure_client()
        else:
            try:
                kube_config.load_incluster_config()
            except ConfigException:
                kube_config.load_kube_config()
        client = ApiClient()
        return client
```